### PR TITLE
Implemented appointment sheet #3785

### PR DESCRIPTION
**Pull request by stgallagher**
### Summary

Appointment sheet that lists a variety of useful information for administering an event.
##### Actions
- Linked the appointment sheet to an icon next to a participant's name in the scheduled activities tab of the welcome page.
- Added a controller, model and view for the appointment sheet
- Placed as much of the logic to derive the information necessary for the sheet in the model. There are a few variables(Event time, Event date, Data Collector name) that I had to do straight from the controller since they are derived from the PSC scheduled activities report. I am looking into ways of migrating these to the model.
##### Comments
- There are some fields in the appointment sheet design mock-up that I couldn't get given that their source objects don't exist prior to an event's execution. Examples are Contact Type, Contact Comments. 
- There are some other fields where the source data doesn't exist or would require a fair amount of work to create, e.g. Pregnancy Length, Communication Preferences, Visit Coordinator name.
#### Design Mock-up

Screen Shot 2013-04-10 at 8 58 37 AM
#### Implementation Screenshot

Screen Shot 2013-04-12 at 2 06 06 PM
#### The printed PDF version (in Firefox) (Admittedly needs work)

Screen Shot 2013-04-12 at 2 24 31 PM

### CODE REVIEW

**Comment by rsutphin** [comment]
The implementation screenshot shows a person ID (XXXX-XXXX-XXXX), not a participant ID (XXX-XX-XXXX) under the participant's name.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 1]
```diff
@@ -0,0 +1,70 @@
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss:0]: These values should be derived from existing Cases metrics, not duplicated here.

File: app/models/event.rb [hunk 2]
```diff
@@ -323,14 +322,6 @@ def strip_time_whitespace
...
   private :strip_time_whitespace
 
-  # Set event start time to now if blank and start date exists
-  def set_start_time
-    if self.event_start_time.blank? && !self.event_start_date.blank?
-      self.event_start_time = Time.now
-    end
-  end
-  private :set_start_time
-
```
Comment (rsutphin) [inline app/models/event.rb:333]: Why was this removed? It seems like a change with far greater impact than adding a new view implies.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+#next-event {
+  margin-top: 1em;
+  margin-top: 1em;
+  margin-left: 3em;
+  font-size: 130%;
+}
+
+.column-container {
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss]: At least based on this, you're using the `@column` mixin in something that isn't a container.  That might work, but it isn't intended usage.

File: app/models/appointment_sheet.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+# Schema version: 20130409233256
+#
+# Table name: appointment_sheets
+#
+#  created_at :datetime
+#  id         :integer          not null, primary key
+#  updated_at :datetime
+#
+
+class AppointmentSheet < ActiveRecord::Base
```
Comment (rsutphin) [inline app/models/appointment_sheet.rb]: You can have a model that isn't an AR model. Since there's nothing persisted in this feature (AFAICT), you don't need to descend from AR::Base.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  font-size: 120%;
+
+}
+
+.large-and-bolded {
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss]: This class name describes a visual effect; it doesn't describe what the class members _are_.  If you're using this class on something that is supposed to be a heading, use the appropriate heading element and develop a style with an appropriate selector.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  font-size: 110%;
+  font-weight: bold;
+}
+
+.column-one {
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss]: Same comments as those on `large-and-bolded`.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  @include column(8);
+  @include colborder();
+  margin-top: 1em;
+  margin-left: 5em;
+}
+
+.column-two {
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss]: Same comments as those on `large-and-bolded`.

File: app/controllers/welcome_controller.rb [hunk 8]
```diff
@@ -122,8 +122,8 @@ def destroy_participant_and_redirect(participant, resp, destroy_person = true)
     end
 
     def get_scheduled_activities_report(options = {})
-      @start_date = 1.day.ago.to_date.to_s
-      @end_date   = params[:end_date] || 6.weeks.from_now.to_date.to_s
+      @start_date = Date.new(2013,4,8)#Date.today.beginning_of_week
```
Comment (hannahwhy) [inline app/controllers/welcome_controller.rb]: Why is this date hardcoded?

File: app/models/appointment_sheet.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    participant_consents(@person)
+  end
+
+  def child_consents
+    consents = []
+    @person.participant.children.each do |child|
+      consents << participant_consents(child)
+    end
+    consents
+  end
```
Comment (hannahwhy) [inline app/models/appointment_sheet.rb:64]:
``` ruby
def child_consents
  @person.participant.children.each_with_object([]) { |c, a| a << participant_consents(c) }
end
```

File: app/models/appointment_sheet.rb [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+  def age(birth_date)
+    today = Date.today
+    total_months = (today.year * 12 + today.month) - (birth_date.year * 12 + birth_date.month)
+    years, months = total_months.divmod(12)
+    strings = [[years, "year"], [months, "month"]].map do |value, unit|
+      case value
+        when 0 then nil
+        when 1 then "#{value} #{unit}"
+        else "#{value} #{unit}s"
+      end
```
Comment (hannahwhy) [inline app/models/appointment_sheet.rb]: ActiveSupport's inflector supplies `String#pluralize`; why not use that?

File: app/models/appointment_sheet.rb [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    @person.participant.children.collect(&:sex).collect(&:display_text)
+  end
+
+  def child_ages
+    ages = []
+    @person.participant.children.each do |child|
+      ages << age(child.person_dob_date)
+    end
+    ages
+  end
```
Comment (hannahwhy) [inline app/models/appointment_sheet.rb:150]: https://github.com/NUBIC/ncs_navigator_core/pull/127/files#r3778970

File: app/models/appointment_sheet.rb (hunk 9 above)
Comment (rsutphin) [inline app/models/appointment_sheet.rb:64]: Or even:

``` ruby
def child_consents
  @person.participant.children.collect { |c| participant_consents(c) }
end
```

File: app/helpers/welcome_helper.rb [hunk 12]
```diff
@@ -16,10 +16,15 @@ def activities(json_response)
...
+  def sort_activities_chronologically_with_blank_times_at_the_bottom_of_list(activities)
+    convert_time_to_integer = Proc.new { |time| time.delete(":").to_i }
+
+    sorted  = activities.select  { |act| act.activity_time }
+                        .sort_by { |act|  convert_time_to_integer.call(act.activity_time) }
+
+    empties = activities.select { |act| act.activity_time.nil?  }
+
+    sorted + empties
   end
```
Comment (hannahwhy) [inline app/helpers/welcome_helper.rb:28]:
``` ruby
def sort(activities)
  activities.sort_by { |a| a.activity_time.blank? ? Float::INFINITY : a.activity_time.delete(':').to_i }
end
```

File: app/views/appointment_sheets/show.html.haml [hunk 13]
```diff
@@ -0,0 +1,66 @@
+%h3#appointment-sheet APPOINTMENT SHEET
+
+.column-container
+  .column-one
+    .large-and-bolded
+      = "#{@sheet.event_type}"
```
Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml]: String interpolation isn't required here.

File: app/models/appointment_sheet.rb (hunk 9 above)
Comment (hannahwhy) [inline app/models/appointment_sheet.rb:64]: Looking at this as a mapping, I'm now wondering if the result ought to be flattened, or if it's intentionally an array of arrays.

File: app/helpers/welcome_helper.rb [hunk 14]
```diff
@@ -16,10 +16,15 @@ def activities(json_response)
     result
   end
 
-  def place_activities_with_blank_times_at_the_bottom_of_list(activities)
-    activities.sort{ |a1, a2|( a1.send(:activity_time) && a2.send(:activity_time) ) ?
-                               a1.send(:activity_time) <=> a2.send(:activity_time) :
-                             ( a1.send(:activity_time) ? -1 : 1 ) }
+  def sort_activities_chronologically_with_blank_times_at_the_bottom_of_list(activities)
+    convert_time_to_integer = Proc.new { |time| time.delete(":").to_i }
```
Comment (rsutphin) [inline app/helpers/welcome_helper.rb:20]: These are zero-padded 24-hour time strings, right? You shouldn't need to convert them to integers to sort them.

File: app/models/event.rb (hunk 2 above)
Comment (stgallagher) [inline app/models/event.rb:333]: It was before I knew that event_start_time != scheduled event time. I talked to Paul and he said it was an out-dated means of setting event_start_time and that they are set by PSC now. I don't need this change anymore since I don't use event_start_time. I will reverse the removal.

File: app/models/appointment_sheet.rb (hunk 4 above)
Comment (stgallagher) [inline app/models/appointment_sheet.rb]: Understood. Will change.

File: app/controllers/welcome_controller.rb (hunk 8 above)
Comment (stgallagher) [inline app/controllers/welcome_controller.rb]: The appointment sheet comes from the scheduled activities page which (according to the design mock-up) is scaled for whatever current week, which changes dynamically (its the commented out code) but I had a series of test subjects assigned to dates so, to keep from having to update their dates, I hard-coded these dates. I do need to change this before merging.

File: app/models/appointment_sheet.rb (hunk 9 above)
Comment (stgallagher) [inline app/models/appointment_sheet.rb:64]: There could be several children and they can each have several consents, so  I think conceptually a two-dimensional array is a good fit.

File: app/models/appointment_sheet.rb (hunk 10 above)
Comment (stgallagher) [inline app/models/appointment_sheet.rb]: Good idea. Thanks.

File: app/views/appointment_sheets/show.html.haml (hunk 13 above)
Comment (stgallagher) [inline app/views/appointment_sheets/show.html.haml]: Good point. Will change.

File: app/helpers/welcome_helper.rb (hunk 14 above)
Comment (stgallagher) [inline app/helpers/welcome_helper.rb:20]: They are _not_ zero-padded, which created the need to do extra stuff for their sorting.

**Comment by stgallagher** [comment]
Ready for Review.

### Summary

Responded to the issues brought up in the first round of code review. Added Cucumber integration tests. 

##### Actions
- Changed erroneous `person_id` to `participant_id`
- Restored `#set_start_time` to `Event` after its unnecessary deletion
- Removed improper CSS inheritence
- Made CSS selectors more semantic
- Date range for scheduled activities now configurable through URL parameters
- Restored default date range to 6 weeks
- Refactored `AppointmentSheet#participant_consents` and `AppointmentSheet#age`
- Adapted methods to handle `nil` conditions more appropriately
- Added integration tests for viewing upcoming appointments and viewing appointment sheets

File: spec/models/address_spec.rb [hunk 15]
```diff
@@ -53,17 +53,18 @@
   end
 
   it "should describe itself" do
-    addr.to_s.should == "#{addr.state}"
+    state_name = "#{addr.state}".downcase.capitalize
```
Comment (rsutphin) [inline spec/models/address_spec.rb]: Expectations should be literal and concrete as much as possible. This just recapitulates what's in the code.

File: app/models/address.rb [hunk 16]
```diff
@@ -98,7 +98,7 @@ def to_s(separator = ' ')
     addr << address_two
     addr << unit
     city_state << city
-    city_state << state.to_s if state && state.local_code != -4
+    city_state << state.to_s.downcase.capitalize if state && state.local_code != -4
```
Comment (rsutphin) [inline app/models/address.rb]: Missed this the first time around: there exist states which have more than one word in their names. You probably want `titleize` here. Note also that `capitalize` and `titleize` both take care of changing the case for all the letters in the string. No need to `downcase` first.

File: db/schema.rb [hunk 17]
```diff
@@ -79,6 +79,11 @@
   add_index "answers", ["question_id"], :name => "idx_answers_question_id"
   add_index "answers", ["reference_identifier"], :name => "idx_answers_reference_identifier"
 
+  create_table "appointment_sheets", :force => true do |t|
+    t.datetime "created_at"
+    t.datetime "updated_at"
+  end
+
```
Comment (rsutphin) [inline db/schema.rb]: This is still here.

File: db/migrate/20130409233256_create_appointment_sheets.rb [hunk 18]
```diff
@@ -0,0 +1,8 @@
+class CreateAppointmentSheets < ActiveRecord::Migration
+  def change
+    create_table :appointment_sheets do |t|
+
+      t.timestamps
+    end
+  end
+end
```
Comment (rsutphin) [inline db/migrate/20130409233256_create_appointment_sheets.rb]: This is still here.

File: db/schema.rb (hunk 17 above)
Comment (stgallagher) [inline db/schema.rb]: Forgot that one. Deleting it and the migration file. Thanks.

File: app/assets/stylesheets/partials/_appointment_sheet.scss [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  margin-top: 1em;
+}
+
+#consents {
+  margin-left: 1em;
+}
+
+#children {
+  margin-top: 1em;
+}
```
Comment (hannahwhy) [inline app/assets/stylesheets/partials/_appointment_sheet.scss:34]: Common style motifs (and this is one) should be factored out into SCSS mixins.  The mixin name, like the style name, should reflect the meaning of the elements to which is applies.

http://sass-lang.com/docs/yardoc/file.SASS_REFERENCE.html#mixins

File: features/appointment_sheet/viewing_an_appointment_sheet.feature [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    Then I should see the the "Environmental Tap Water Pharmaceuticals Technician Collect Instrument" as to be conducted
+    Then I should see the the "Environmental Vacuum Bag Dust Technician Collect Instrument" as to be conducted
+    Then I should see the the "Environmental Vacuum Bag Dust (VBD) Participant Collect SAQ Specification" as to be conducted
+    Then I should see the the "Environmental Sample Kit Distribution Instrument" as to be conducted
+    Then I should see the the "Environmental Sample Kit Distribution Instrument" as to be conducted
+    Then I should see the the "Tracing Module" as to be conducted
+    Then I should see the the "Birth Visit Information Sheet" as to be conducted
+    Then I should see the the "Pregnancy Visit 1 Information Sheet" as to be conducted
+    Then I should see the the "Pregnancy Health Care Log" as to be conducted
+    Then I should the next event as "Pregnancy Visit 2"
```
Comment (hannahwhy) [inline features/appointment_sheet/viewing_an_appointment_sheet.feature]: This is the sort of thing that tables are very good at.

File: features/appointment_sheet/viewing_upcoming_appointments.feature [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+      | person/last_name  | Folgers             |
+      | person/person_id  | fy84-54sf-wc3s      |
+      | p_id              | abc-128-wxyz        |
+
+  @javascript
+  Scenario: The schedule lists my upcoming contacts for the next week
+    Given an authenticated user
+    When I view my scheduled events for 2013-04-08 to 2013-04-13
+    Then I should see "Dawn Davidson" scheduled for "Monday April 8" for a "Pregnancy Visit 1" event at "2:45 PM"
+    Then I should see "Elaine Edwinson" scheduled for "Wednesday April 10" for a "Pbs Participant Eligibility Screening" event at "9:30 AM"
```
Comment (hannahwhy) [inline features/appointment_sheet/viewing_upcoming_appointments.feature]: When chaining `Then`s (if you must -- I prefer that each scenario have a single `Then`, though sometimes it's neither convenient nor readable) use `And`.  The same applies for Given and When.

Keep in mind that a scenario is meant to be read _by a human_.  (So is source code, for that matter.)  Ideally, that human is a domain expert with no knowledge of the implementation.  It is therefore _very_ important that your scenario read well.

File: features/step_definitions/appointment_sheet_steps.rb [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+  @participant.save!
+end
+
+Given /^speaks english$/ do
+  @person.update_attribute(:language_code, 1)
+end
+
+Given /^is a boy$/ do
+  @child_person.update_attribute(:sex_code, 1)
+end
```
Comment (hannahwhy) [inline features/step_definitions/appointment_sheet_steps.rb:140]: This step isn't appointment-sheet specific.

Where possible, avoid grouping steps by scenario.  Strive to reuse existing steps, extending them if doing so is not too difficult.  (It's a judgment call; there is no hard-and-fast criteria.)  Building out specific steps for scenarios has a nasty habit of developing into scenarios that use slightly different steps to express the same actions and preconditions, which significantly impairs maintainability.

File: features/step_definitions/appointment_sheet_steps.rb [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+  step "I should see #{name}"
+  step "I should see #{date}"
+  step "I should see #{event}"
+end
+
+Then /^I should see ("[^"]*") scheduled for ("[^"]*") for a ("[^"]*") event at ("[^"]*")$/ do |name, date, event, time|
+  step "I should see #{name}"
+  step "I should see #{date}"
+  step "I should see #{event}"
+  step "I should see #{time}"
```
Comment (hannahwhy) [inline features/step_definitions/appointment_sheet_steps.rb]: Use `steps`.

``` ruby
Then /^I should see ("[^"]*") scheduled for ("[^"]*") for a ("[^"]*") event at ("[^"]*")$/ do |name, date, event, time|
  steps %Q{
    Then I should see "#{name}"
    And I should see "#{date}"
 }
end
```

File: features/step_definitions/appointment_sheet_steps.rb [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+end
+
+Then /^I should see scheduled event ("[^"]*")$/ do |event|
```
Comment (hannahwhy) [inline features/step_definitions/appointment_sheet_steps.rb]: Avoid `should`.  It's a noise word.

In the context of Cucumber (and RSpec), "Then I see" means the same thing as "Then I should see".  A violation of either expectation will be logged.

File: features/support/fakeweb.rb [hunk 25]
```diff
@@ -8,7 +8,10 @@
 FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/registered_with_psc\/schedules.json$/,
                     :body => "#{Rails.root}/features/fixtures/fakeweb/registered_with_psc_schedule.json", :status => ["200", "OK"], :content_type => "application/json")
 
-FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/((?!registered_with_psc))/, :body => "Unknown", :status => ["401", "Unknown"])
+FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/xfda-a78s-h83b\/schedules.json$/,
+                    :body => "#{Rails.root}/features/fixtures/fakeweb/build_activity_plan_response_for_appointment_sheet.json", :status => ["200", "OK"], :content_type => "application/json")
+
+#FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/((?!registered_with_psc))/, :body => "Unknown", :status => ["401", "Unknown"])
```
Comment (hannahwhy) [inline features/support/fakeweb.rb]: If this isn't needed, eliminate it.

**Comment by stgallagher** [comment]
Added due date and birth date to the appointment sheet per Sean's request. Fixed a few of the noted issues:
- deleted migration and schema changes
- deleted bad test from address spec

File: app/models/appointment_sheet.rb [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+  def cell_phone
+    phone = Telephone.where(:person_id => @person.id,
+                            :phone_type_code => 3,
+                            :phone_rank_code => 1).first
+    return nil if phone.nil?
+    if phone.phone_nbr =~ /-/
+      return phone.phone_nbr
+    end
+    phone.phone_nbr.insert(-5, '-').insert(-9, '-') if phone
+  end
```
Comment (hannahwhy) [inline app/models/appointment_sheet.rb:36]: The formatting code here seems like it belongs on Telephone.  (It's formatting, right?  It looks like you're asking for a phone number in XXX-XXX-XXXX format.)

I suggest this:

``` ruby
phone.to_s
```

Or, if `#to_s` is already in use and can't be changed, something similar.  `#dashed` is one possibility.

File: app/views/appointment_sheets/show.html.haml [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+        - @sheet.children.each_index do |child_index|
+          Child:
+          %span.name
+            #{@sheet.child_names[child_index]}
+          .child_attributes
+            - if @sheet.child_birth_dates[child_index]
+              = "Birth date: #{@sheet.child_birth_dates[child_index]}"
+            - else
+              .not-found
+                Unknown birth_date
```
Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml]: All of these if-elses can be handled with `||`, i.e.

``` ruby
"Birth date: #{birth_date}" || "Unknown birth date"
```

This is a good place to introduce helpers.

Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml]: Also, it's "birth date", not "birth_date".

File: app/views/appointment_sheets/show.html.haml [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+  - activities_for_event = @participant_activity_plan.scheduled_activities_for_event(pending_event) if pending_event
+  #instrument-list
+    - if @participant_activity_plan && pending_event
+      %span#to-be-conducted
+        To be conducted:
+      .activities
+        - activities_for_event.each do |a|
+          .activity
+            %span{ :title => a.current_state}
+              &nbsp;
```
Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml:129]: Express this in styles, not via non-breaking spaces.

File: app/views/appointment_sheets/show.html.haml [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+                Unknown age
+          Consents:
+          - if @sheet.child_consents.empty?
+            No consents found for children
+          - else
+            - if @sheet.child_consents[child_index].nil?
+              .not-found
+                No consents found for this child
+            - else
+              - @sheet.child_consents[child_index].uniq.each do |consent|
```
Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml]: Why would you be getting duplicate consent records?

File: app/views/appointment_sheets/show.html.haml [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+      - else
+        - @sheet.mothers_consents.uniq.each do |consent|
+          #consents
+            = consent
+      - if @sheet.children.empty?
+        .not-found
+          No children found
+      - else
+        #children
+        - @sheet.children.each_index do |child_index|
```
Comment (hannahwhy) [inline app/views/appointment_sheets/show.html.haml:69]: Instead of this, you can avoid having to propagate `child_index` everywhere by building an intermediate object for each child and iterating over that.  That implies that each object in `@sheet.children` would contain all information for that child; you wouldn't have to dereference anything.

**Comment by stgallagher** [comment]
Ready for review.

### Summary

Responding to 2nd round of feedback

##### Actions
- Removed `#uniq` from consent listing. This method was covering up a duplication bug that I addressed by changing `#participant_consents`
- Moved phone formatting logic to `Telephone` as `#dashed`
- Used `#titleize`for state string output
- Removed should from features
- Wrapped multi-step features
- Moved non-specific steps into `common_steps`
- Changed multiple successive 'Then's to 'And's
- Fixed `birth_date` typo 

##### Deferring
- The construction of an intermediate child object : this is actually a different (better) approach but involves a good deal of reworking my current approach, which I think works, just isn't as conceptually sound
- Extracting if-else in the view to helper methods: if I have time to return to this I can do it, but I don't think right now it's justified.

##### Remaining Issues
- Some of the Cucumber tests fail when I run `rake cucumber:all` locally 
- I need to determine if the FakeWeb route I commented out is used and where (it's signature was covering up a route I needed to fake out PSC#build_activity_plan

File: app/controllers/appointment_sheets_controller.rb [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+    @event_date = event_date
+    @event_time = event_time
+  end
+
+  private
+
+  def event_time
+    event_activity = @participant_activity_plan.scheduled_activities.find do |sa|
+      sa.date == params[:date] &&
+      sa.event == @sheet.event_type.downcase.gsub(/ /,'_') &&
```
Comment (hannahwhy) [inline app/controllers/appointment_sheets_controller.rb]: A regexp isn't necessary for this; `String#gsub` handles strings just fine.  You can also use `String#tr`:

``` ruby
sa.event == @sheet.event_type.downcase.tr(' ', '_')
```

File: app/helpers/application_helper.rb [hunk 32]
```diff
@@ -135,7 +135,7 @@ def staff_list
   end
 
   def build_staff_list
-    users = Aker.authority.find_users
+    users = Aker.configuration.authorities.find { |auth| NcsNavigator::Authorization::Core::Authority === auth }.find_users
```
Comment (hannahwhy) [inline app/helpers/application_helper.rb:138]: Why not just instantiate an appropriate Authority for this purpose?

(I'd recommend moving this out of `ApplicationHelper`, too)

Comment (rsutphin) [inline app/helpers/application_helper.rb:138]: > Why not just instantiate an appropriate Authority for this purpose?

Because if the authority requires configuration, this will get an instance which is known to be appropriately configured for the application and environment.

File: app/controllers/appointment_sheets_controller.rb (hunk 31 above)
Comment (stgallagher) [inline app/controllers/appointment_sheets_controller.rb]: `#tr` reflects more specifically what I intended. Didn't know about that method. I've changed it in the next set of revisions.

**Comment by stgallagher** [comment]
Print sheet is getting better. If anyone knows how to specify "always scale to a single page" let me know please.
Screen Shot 2013-04-18 at 4 10 26 PM

**Comment by stgallagher** [comment]
Ready for review.

### Summary

Added last contact comment. Formatted print view. Responded to 3rd round of feedback.

##### Actions
- Getting the last comment from all the contacts associated with the person and all the contacts associated with that person's events
- Added stylesheet to expressly handle printing the appointment sheet
- Used #tr instead of #gsub in AppointmentSheetController#event_time since it's more reflective of my intent
- Tweeked the view
- Inserting lines of contingency code to handle potential nil source objects

Final (for now) print style - (If anyone knows who to get rid of the annoying URL under "Next Event" I'd appreciate it)
Screen Shot 2013-04-19 at 2 09 02 PM

**Comment by pfriedman** [comment]
Merged PR into `develop`
